src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -5876,6 +5878,20 @@ private ScanTypeArgumentListKind ScanTypeArgumentList(NameOptions options)
 
             if (this.CurrentToken.Kind != SyntaxKind.GreaterThanToken)
             {
+                // for identifiers we assume that there could be a missing > token
+                // for example, we have reached C in X<A, B C
+
+                // for tuples, we do not expect direct invocation right after the parenthesis
+                // an example is X<(string, string)(), where we imply a missing token between )(
+                // as the user probably wants to invoke X by X<(string, string)>()
+
+                if (this.CurrentToken.Kind is SyntaxKind.IdentifierToken ||
+                    (lastScannedType is ScanTypeFlags.TupleType && this.CurrentToken.Kind is SyntaxKind.OpenParenToken))
+                {
+                    greaterThanToken = this.EatToken(SyntaxKind.GreaterThanToken);
+                    return result;
+                }
```

Comment:
i woudl break these into separate checks.  just to make it easy to undrstand what's going on, and to be clear in the future about how to add new cases.  i would also doc that new cases are allowed to be added here as long as it is clear that whatever code it detects must be an error case in some way, so this is just about better error recovery, not about parsing correct constructs correctly.